File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/agent/AgentParseMsgTest.java
package org.hyperledger.indy.sdk.agent;

import org.hyperledger.indy.sdk.IndyIntegrationTestWithPoolAndSingleWallet;
import org.hyperledger.indy.sdk.InvalidStructureException;
import org.hyperledger.indy.sdk.crypto.Crypto;
import org.hyperledger.indy.sdk.crypto.CryptoJSONParameters;
import org.hyperledger.indy.sdk.signus.Signus;
import org.hyperledger.indy.sdk.signus.SignusResults;
import org.hyperledger.indy.sdk.wallet.WalletValueNotFoundException;
import org.junit.Test;

import java.util.Arrays;
import java.util.concurrent.ExecutionException;

import static org.hamcrest.CoreMatchers.isA;
import static org.junit.Assert.*;


public class AgentParseMsgTest extends IndyIntegrationTestWithPoolAndSingleWallet {
	@Test
	public void testParseMsgWorksForAuthenticatedMessage() throws Exception {
		String paramJson = new CryptoJSONParameters.CreateKeyJSONParameter(MY1_SEED, null).toJson();
		String senderVk = Crypto.createKey(wallet, paramJson).get();

		paramJson = new CryptoJSONParameters.CreateKeyJSONParameter(MY2_SEED, null).toJson();
		String recipientVk = Crypto.createKey(wallet, paramJson).get();

		byte[] encryptedMsg = Agent.prepMsg(wallet, senderVk, recipientVk, MESSAGE).get();
		AgentResults.ParseMsgResult parseResult = Agent.parseMsg(wallet, recipientVk, encryptedMsg).get();
		assertEquals(senderVk, parseResult.getSenderKey());
		assertTrue(Arrays.equals(MESSAGE, parseResult.getMessage()));
	}

	@Test
	public void testParseMsgWorksForAnonymousMessage() throws Exception {
		String paramJson = new CryptoJSONParameters.CreateKeyJSONParameter(MY2_SEED, null).toJson();
		String recipientVk = Crypto.createKey(wallet, paramJson).get();

		byte[] encryptedMsg = Agent.prepAnonymousMsg(recipientVk, MESSAGE).get();
		AgentResults.ParseMsgResult parseResult = Agent.parseMsg(wallet, recipientVk, encryptedMsg).get();
		assertNull(parseResult.getSenderKey());
		assertTrue(Arrays.equals(MESSAGE, parseResult.getMessage()));
	}

	@Test
	public void testParseMsgWorksForInvalidAuthenticatedMessage() throws Exception {
		SignusResults.CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(wallet, "{}").get();
		String recipientDid = result.getDid();
		String recipientVk = result.getVerkey();

		String identityJson = String.format(IDENTITY_JSON_TEMPLATE, recipientDid, recipientVk);
		Signus.storeTheirDid(wallet, identityJson).get();

		String msg = String.format("{\"auth\":true,\"nonсe\":\"Th7MpTaRZVRYnPiabds81Y12\",\"sender\":\"%s\",\"msg\":\"unencrypted message\"}", VERKEY);
		byte[] encryptedMsg = Signus.encryptSealed(wallet, pool, recipientDid, msg.getBytes()).get();

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		Agent.parseMsg(wallet, recipientVk, encryptedMsg).get();
	}

	@Test
	public void testParseMsgWorksForInvalidAnonymousMessage() throws Exception {
		String recipientVk = Crypto.createKey(wallet, "{}").get();

		String msg = "unencrypted message";

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		Agent.parseMsg(wallet, recipientVk, msg.getBytes()).get();
	}

	@Test
	public void testParseMsgWorksForUnknownRecipientVk() throws Exception {
		byte[] encryptedMsg = Agent.prepAnonymousMsg(VERKEY, MESSAGE).get();

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(WalletValueNotFoundException.class));

		Agent.parseMsg(wallet, VERKEY, encryptedMsg).get();
	}
}

File: wrappers/dotnet/indy-sdk-dotnet-test/AgentTests/AgentParseMsgTest.cs
using Hyperledger.Indy.AgentApi;
using Hyperledger.Indy.CryptoApi;
using Hyperledger.Indy.SignusApi;
using Hyperledger.Indy.WalletApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.AgentTests
{
    [TestClass]
    public class AgentParseMsgTest : IndyIntegrationTestWithPoolAndSingleWallet
    {
        const string _keyJsonTemplate = "{{\"seed\":\"{0}\"}}";

        [TestMethod]
        public async Task TestParseMsgWorksForAuthenticatedMessage()
        {
            var paramJson = string.Format(_keyJsonTemplate, MY1_SEED);
            var senderVk = await Crypto.CreateKeyAsync(wallet, paramJson);

            paramJson = string.Format(_keyJsonTemplate, MY2_SEED);
            var recipientVk = await Crypto.CreateKeyAsync(wallet, paramJson);

            var encryptedMsg = await Agent.PrepMsgAsync(wallet, senderVk, recipientVk, MESSAGE);
            var parseResult = await Agent.ParseMsgAsync(wallet, recipientVk, encryptedMsg);

            Assert.AreEqual(senderVk, parseResult.SenderKey);
            Assert.IsTrue(MESSAGE.SequenceEqual(parseResult.Msg));
        }

        [TestMethod]
        public async Task TestParseMsgWorksForAnonymousMessage()
        {
            var paramJson = string.Format(_keyJsonTemplate, MY2_SEED);
            var recipientVk = await Crypto.CreateKeyAsync(wallet, paramJson);

            var encryptedMsg = await Agent.PrepAnonymousMsgAsync(recipientVk, MESSAGE);
            var parseResult = await Agent.ParseMsgAsync(wallet, recipientVk, encryptedMsg);

            Assert.IsNull(parseResult.SenderKey);
            Assert.IsTrue(MESSAGE.SequenceEqual(parseResult.Msg));
        }

        [TestMethod]
        public async Task TestParseMsgWorksForInvalidAuthenticatedMessage()
        {
            var result = await Signus.CreateAndStoreMyDidAsync(wallet, "{}");
            var recipientDid = result.Did;
            var recipientVk = result.VerKey;

            var identityJson = string.Format(IDENTITY_JSON_TEMPLATE, recipientDid, recipientVk);
            await Signus.StoreTheirDidAsync(wallet, identityJson);

            var msg = string.Format("{{\"auth\":true,\"nonсe\":\"Th7MpTaRZVRYnPiabds81Y12\",\"sender\":\"{0}\",\"msg\":\"unencrypted message\"}}", VERKEY);
            var encryptedMsg = await Signus.EncryptSealedAsync(wallet, pool, recipientDid, Encoding.UTF8.GetBytes(msg));

            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                Agent.ParseMsgAsync(wallet, recipientVk, encryptedMsg)
            );
        }

        [TestMethod]
        public async Task TestParseMsgWorksForInvalidAnonymousMessage()
        {
            var recipientVk = await Crypto.CreateKeyAsync(wallet, "{}");

            var msg = "unencrypted message";
            var encryptedMsg = Encoding.UTF8.GetBytes(msg);

            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                Agent.ParseMsgAsync(wallet, recipientVk, encryptedMsg)
            );
        }

        [TestMethod]
        public async Task TestParseMsgWorksForUnknownRecipientVk()
        {
            var encryptedMsg = await Agent.PrepAnonymousMsgAsync(VERKEY, MESSAGE);

            var ex = await Assert.ThrowsExceptionAsync<WalletValueNotFoundException>(() =>
                Agent.ParseMsgAsync(wallet, VERKEY, encryptedMsg)
            );
        }
    }
}
